sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala

```diff
@@ -227,3 +228,14 @@ object ReuseSubquery extends Rule[SparkPlan] {
     }
   }
 }
+
+object PrepareScans extends Rule[SparkPlan] {
+  def apply(plan: SparkPlan): SparkPlan = {
+    val scans = plan.collect {
+      case scan: BatchScanExec => scan
```

Comment:
It's really tricky to make `BatchScanExec` mutable and update them here. Can we just create new `Scan` instance here and return a new `BatchScanExec` with the new `Scan`?

Reply:
I'd like to avoid that too. I am using `prepare` here as it is idempotent. If we return a new `BatchScanExec`, will that mean the subqueries will be executed twice as the new plan won't be prepared?

Reply:
I made `BatchScanExec` immutable right now.

Reply:
Resolving this one too as it no longer applies.